aspnet/mvc/controllers/routing.rst

```diff
@@ -1,8 +1,140 @@
+
+Routes on controller vs actions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Routes map URLs to actions in controllers. When at the action level, constraints may be applied to the actions to further restrict activity. 
+
+Inheritance
+^^^^^^^^^^^
+
+The ``IRouteConstraint`` interface exposes a ``Match`` method that you can use to create your own route constraints via inheritance, as the code below demonstrates:
+
+.. code-block:: c#
+
+  public class RouteConstraint : IRouteConstraint
+  {
+    public bool Match(HttpContext httpContext, 
+        IRouter route, string routeKey, 
+        IDictionary<string, object> values, 
+        RouteDirection routeDirection)
+        {
+          // code to build constraint
+        }
+  }
+
+Notice that the ``Match`` method contains parameters that you can use to inspect the route and HTTP request.
+  
+Token replacement in route templates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Since maintainability is important, you can use token replacements in route templates. Tokens are substitutions for actual route values at runtime, and you may define them during startup for conventional routes, or in attributes for attribute routes. You may use regular expressions to create route tokens. Consider the following route attribute:
+
+.. code-block:: c#
+
+  [Route("category/{movie}.rss", Order = 0, Name = "MovieCategoryFeed")]
+  public IActionResult MovieRSS(string name)
+
+The above creates a route that directs to the ``MovieRSS`` method, and looks something like the following:
+
+.. code-block:: c#
+
+  /category/GoneWithTheWind.rss 
+
+You may create custom route templates if you can't find any other built-in solution. Generally, it is because you need to dynamically compute a route template. To do so, subclass any route attribute, or implement the ``IApplicationModelConvention`` interface. Keep the code in the custom attribute simple as a limited extensibility point.   
+
+Using Application Model to customize attribute routes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Implement the ``Apply`` method of the ``IApplicationModelConvention`` interface to create an attribute route. For more information on the MVC application model, see the documentation on `ApplicationModel <https://docs.asp.net/en/latest/mvc/controllers/application-model.html>`_.
+
+Action constraints
+------------------
+
+MVC provides a few ways to apply finer control over how it processes routes. For example, you might want to restrict URLs to a particular HTTP verb. Other times you might want to dynamically prepend or append strings to routes, such as "api" for routes that lead to an API action. Or perhaps you're working on a blog engine and want the publish date and SEO information in the URLs. Constraints are great for when you need to modify the default behavior of actions.
```

Comment:
> Other times you might want to dynamically prepend or append strings to routes, such as "api" for routes that lead to an API action. Or perhaps you're working on a blog engine and want the publish date and SEO information in the URLs. Constraints are great for when you need to modify the default behavior of actions.

These aren't descriptions are of what `IActionConstraint` allows. `IActionConstraint` allows you to implement policies that disambiguate between actions. See a sample here: https://github.com/aspnet/Entropy/tree/dev/samples/Mvc.ActionConstraintSample.Web